File: Coding-Patterns/Coding-Patterns-CSharp/CodingPatterns/08_linked_lists/LruCache.cs
// LeetCode 146 - LRU Cache
// Difficulty: Medium
// Pattern: HashMap + Doubly Linked List
//
// Time: O(1) for get and put  Space: O(capacity)

namespace CodingPatterns.LinkedLists;

public class LRUCache
{
    private class Node
    {
        public int Key, Val;
        public DateTime? ExpiresAtUtc; // null = no expiration
        public Node Prev, Next;
        public Node(int k = 0, int v = 0) 
        { 
            Key = k; Val = v; 
        }
    }

    private readonly int _capacity;
    private readonly Dictionary<int, Node> _cache = new();
    private readonly Node _head = new(); // LRU end (dummy)
    private readonly Node _tail = new(); // MRU end (dummy)

    public LRUCache(int capacity)
    {
        _capacity = capacity;
        _head.Next = _tail;
        _tail.Prev = _head;
    }

    private void Remove(Node node)
    {
        node.Prev.Next = node.Next;
        node.Next.Prev = node.Prev;
    }

    private void InsertTail(Node node)
    {
        node.Prev = _tail.Prev;
        node.Next = _tail;
        _tail.Prev.Next = node;
        _tail.Prev = node;
    }

    public int Get(int key)
    {
        if (!_cache.TryGetValue(key, out var node)) 
            return -1;
        if (node.ExpiresAtUtc.HasValue && node.ExpiresAtUtc.Value <= DateTime.UtcNow)
        {
            Remove(node);
            _cache.Remove(key);
            return -1;
        }
        Remove(node);
        InsertTail(node);
        return node.Val;
    }

    public void Put(int key, int value) => PutInternal(key, value, expiresAtUtc: null);

    // Same as Put, but the entry is treated as absent (by Get) once ttl has elapsed,
    // even if it hasn't been evicted by capacity yet.
    public void PutWithTtl(int key, int value, TimeSpan ttl)
    {
        if (ttl <= TimeSpan.Zero)
            throw new ArgumentOutOfRangeException(nameof(ttl));
        PutInternal(key, value, DateTime.UtcNow + ttl);
    }

    private void PutInternal(int key, int value, DateTime? expiresAtUtc)
    {
        if (_cache.ContainsKey(key))
            Remove(_cache[key]);

        var node = new Node(key, value)
        {
                ExpiresAtUtc = expiresAtUtc
        };
        _cache[key] = node;

        InsertTail(node);
        
        if (_cache.Count > _capacity)
        {
            var lru = _head.Next;
            Remove(lru);
            _cache.Remove(lru.Key);
        }
    }

    public static void Run()
    {
        var lru = new LRUCache(2);
        lru.Put(1, 1); lru.Put(2, 2);
        Console.WriteLine(lru.Get(1));  // 1
        lru.Put(3, 3);                  // evicts key 2
        Console.WriteLine(lru.Get(2));  // -1
        lru.Put(4, 4);                  // evicts key 1
        Console.WriteLine(lru.Get(1));  // -1
        Console.WriteLine(lru.Get(3));  // 3
        Console.WriteLine(lru.Get(4));  // 4

        var ttlCache = new LRUCache(2);
        ttlCache.PutWithTtl(5, 5, TimeSpan.FromMilliseconds(50));
        Console.WriteLine(ttlCache.Get(5));  // 5 (not yet expired)
        Thread.Sleep(75);
        Console.WriteLine(ttlCache.Get(5));  // -1 (expired, though never evicted by capacity)
    }
}

File: Coding-Patterns/Coding-Patterns-Python/08_linked_lists/lru_cache.py
# LeetCode 146 - LRU Cache
# Difficulty: Medium
# Pattern: HashMap + Doubly Linked List
#
# Problem: Design a data structure that follows LRU (Least Recently Used)
# cache eviction. Support get(key) and put(key, value) in O(1).
#
# Approach:
#   - HashMap: key -> node for O(1) lookup
#   - Doubly linked list: maintains usage order
#     - Most recently used near tail (dummy)
#     - Least recently used near head (dummy)
#   - On get/put: move accessed node to tail (most recent)
#   - On capacity exceeded: remove node at head (least recent)
#
# Time: O(1) for get and put  Space: O(capacity)

class DLinkedNode:
    def __init__(self, key=0, val=0):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None

class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}  # key -> node

        # Dummy head (LRU end) and tail (MRU end)
        self.head = DLinkedNode()
        self.tail = DLinkedNode()
        self.head.next = self.tail
        self.tail.prev = self.head

    def _remove(self, node: DLinkedNode):
        node.prev.next = node.next
        node.next.prev = node.prev

    def _insert_tail(self, node: DLinkedNode):
        node.prev = self.tail.prev
        node.next = self.tail
        self.tail.prev.next = node
        self.tail.prev = node

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self._remove(node)
        self._insert_tail(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self._remove(self.cache[key])
        node = DLinkedNode(key, value)
        self.cache[key] = node
        self._insert_tail(node)
        if len(self.cache) > self.capacity:
            lru = self.head.next
            self._remove(lru)
            del self.cache[lru.key]


# ---- Tests ----
if __name__ == "__main__":
    lru = LRUCache(2)
    lru.put(1, 1)
    lru.put(2, 2)
    assert lru.get(1) == 1       # returns 1, key 1 now MRU
    lru.put(3, 3)                # evicts key 2 (LRU)
    assert lru.get(2) == -1      # key 2 was evicted
    lru.put(4, 4)                # evicts key 1 (LRU)
    assert lru.get(1) == -1      # key 1 was evicted
    assert lru.get(3) == 3
    assert lru.get(4) == 4
    print("All tests passed.")
